File: 433-minimum-genetic-mutation/minimum-genetic-mutation.cpp
class Solution {
public:
    int minMutation(string startGene, string endGene, vector<string>& bank) {

        unordered_set<string> bankSet(bank.begin(), bank.end());
        unordered_set<string> visited;

        queue<string> que;
        que.push(startGene);
        int level = 0;


        while(!que.empty()) {

            

            int n = que.size();

            while(n--) {

                string curr = que.front();
                que.pop();

                if(curr == endGene) return level;

                for(auto &ch : "ACGT") {

                    for(int i = 0; i < 8; i++) {

                        char c = curr[i];
                        curr[i] = ch;

                        if(bankSet.find(curr) != bankSet.end() && visited.find(curr) == visited.end()) {
                            que.push(curr);
                            visited.insert(curr);
                        }

                        curr[i] = c;
                    }
                }

            }

            level++;
        }

        return -1;


        
    }
};

File: 433-minimum-genetic-mutation/minimum-genetic-mutation.js
/**
 * @param {string} startGene
 * @param {string} endGene
 * @param {string[]} bank
 * @return {number}
 */
var minMutation = function(startGene, endGene, bank) {

    const visited = new Set();
    const bankSet = new Set(bank);
    visited.add(startGene);

    return BFS();

    function BFS() {

        const queue = new Array();

        queue.push(startGene);

        let level = 0;

        while(queue.length) {

            let n = queue.length;

            while(n--) {

                let curr = queue.shift();

                if(curr == endGene) return level;

                curr = curr.split("");

                for(const ch of "ACGT") {

                    for(let i = 0; i < 8; i++) {
                        let c = curr[i];
                        curr[i] = ch;
                        let t = curr.join("");
                        curr[i] = c;

                        if(bankSet.has(t) && !visited.has(t)) {
                            queue.push(t);
                            visited.add(t);
                        
                        }
                       
                    }
                }

            }

            level++;
        }

        return -1;

    }
    
};